### PR TITLE
Add Digilent Genesys2 board

### PR DESCRIPTION
**Pull request by peteut**
The `Blinky` test works fine on my board.

### CODE REVIEW

File: nmigen_boards/genesys2.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+    resources = [
+        Resource("sysclk", 0, DiffPairs(p="AD12 ", n="AD11", dir="i"),
+                 Clock(200e6), Attrs(IOSTANDARD="LVDS")),
+        Resource("cpu_reset", 0, PinsN("R19", dir="i"),
+                 Attrs(IOSTANDARD="LVCMOS33")),
+        *LEDResources(pins="T28 V19 U30 U29 V20 V26 W24 W23",
+                      attrs=Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("fan", 0,
+                 Subsignal("pwm", Pins("W19", dir="o")),
+                 Subsignal("tack", Pins("V21", dir="i")),
```
[the remark this answers is not in the document]
Comment (peteut) [reply nmigen_boards/genesys2.py:27]: Should spell `tach`, each rotation generates 4 pulses.

File: nmigen_boards/genesys2.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+class Genesys2Platform(Xilinx7SeriesPlatform):
+    device = "xc7k325t"
+    package = "ffg900"
+    speed = "2"
+    default_clk = "sysclk"
+    default_rst = "cpu_reset"
+    resources = [
+        Resource("sysclk", 0, DiffPairs(p="AD12 ", n="AD11", dir="i"),
+                 Clock(200e6), Attrs(IOSTANDARD="LVDS")),
+        Resource("cpu_reset", 0, PinsN("R19", dir="i"),
```
[the remark this answers is not in the document]
Comment (peteut) [reply nmigen_boards/genesys2.py:21]: Indedd, the label on the board is just `RESET`, the refence manual label is `CPU_RESET`. `rst` is just fine.

File: nmigen_boards/genesys2.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+                 Attrs(IOSTANDARD="LVCMOS33")),
+        *LEDResources(pins="T28 V19 U30 U29 V20 V26 W24 W23",
+                      attrs=Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("fan", 0,
+                 Subsignal("pwm", Pins("W19", dir="o")),
+                 Subsignal("tack", Pins("V21", dir="i")),
+                 Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("serial", 0,
```
[the remark this answers is not in the document]
Comment (peteut) [reply nmigen_boards/genesys2.py:29]: Okay.

File: nmigen_boards/genesys2.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+                   "ha10_n": "N22",
+                   "ha17_cc_p": "C25",
+                   "ha17_cc_n": "B25",
+                   "ha21_p": "G28",
+                   "ha21_n": "F28",
+                   "ha23_p": "G18",
+                   "ha23_n": "F18"})]
+
+    def __init__(self, vadj: float = 2.5) -> None:
+        if vadj not in [1.2, 1.8, 2.5, 3.3]:
```
[the remark this answers is not in the document]
Comment (peteut) [reply nmigen_boards/genesys2.py:325]: I see, should I remove the annotations as well?

File: nmigen_boards/genesys2.py [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+                           Attrs(IOSTANDARD="LVCMOS33"))),
+        Resource("vga", 0,
+                 Subsignal("r", Pins("AK25 AG25 AH25 AK24 AJ24", dir="o")),
+                 Subsignal("g", Pins("AJ23 AJ22 AH22 AK21 AJ21 AK23",
+                                     dir="o")),
+                 Subsignal("b", Pins("AH20 AG20 AF21 AK20 AG22", dir="o")),
+                 Subsignal("hsync", PinsN("AF20", dir="o")),
+                 Subsignal("vsync", PinsN("AG23", dir="o")),
+                 Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("mmc", 0,
```
[the remark this answers is not in the document]
Comment (peteut) [reply nmigen_boards/genesys2.py:125]: I see, `SDCardResource` it is.

[the remark this answers is not in the document]
Comment (peteut) [reply nmigen_boards/genesys2.py:125]: I'll add `sd_card_reset` as in `Nexys4DDRPlatform`, as `SDCardResource` does not cover it.

[the remark this answers is not in the document]
Comment (whitequark) [reply nmigen_boards/genesys2.py:125]: The reset physically removes power from the SD card, right? That should be separate, yes.

File: nmigen_boards/genesys2.py [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+                   "ha21_n": "F28",
+                   "ha23_p": "G18",
+                   "ha23_n": "F18"})]
+
+    def __init__(self, vadj: float = 2.5) -> None:
+        if vadj not in [1.2, 1.8, 2.5, 3.3]:
+            raise ValueError("Unknown vadj of {!r}".format(vadj))
+
+        lvcmos_vadj = Attrs(IOSTANDARD="LVCMOS{}".format(int(vadj * 10)))
+        self.resources = Genesys2Platform.resources + [
```
[the remark this answers is not in the document]
Comment (peteut) [reply nmigen_boards/genesys2.py:329]: I see, adding a method `bank15_16_17_iostandard(self)` as in `AtlysPlatform`.

File: nmigen_boards/genesys2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (peteut) [reply nmigen_boards/genesys2.py:125]: That's correct, handle it as in `Nexys4DDRPlatform`.

File: nmigen_boards/genesys2.py [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+
+
+class Genesys2Platform(Xilinx7SeriesPlatform):
+    device = "xc7k325t"
+    package = "ffg900"
+    speed = "2"
+    default_clk = "sysclk"
+    default_rst = "cpu_reset"
+    resources = [
+        Resource("sysclk", 0, DiffPairs(p="AD12 ", n="AD11", dir="i"),
```
[the remark this answers is not in the document]
Comment (peteut) [reply nmigen_boards/genesys2.py]: Done.

File: nmigen_boards/genesys2.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (peteut) [reply nmigen_boards/genesys2.py]: Typo fixed.

File: nmigen_boards/genesys2.py [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+                "e": "C19",
+                "n": "B19",
+                "s": "M19",
+                "c": "E18"}, attrs=Attrs(IOSTANDARD=bank15_16_17_iostandard)),
+        *SwitchResources(pins="G19 G25 H24 K19 N19 P19",
+                         attrs=Attrs(IOSTANDARD=bank15_16_17_iostandard)),
+        *SwitchResources(pins={
+                6: "P26",
+                7: "P27"}, attrs=Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("clk", 0, DiffPairs(p="AD12 ", n="AD11", dir="i"),
```
Comment (whitequark) [inline nmigen_boards/genesys2.py:44]: By convention, clock and reset resources should be at the top.

File: nmigen_boards/genesys2.py [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+                 Subsignal("pwm", Pins("W19", dir="o")),
+                 Subsignal("tach", Pins("V21", dir="i")),
+                 Attrs(IOSTANDARD="LVCMOS33")),
+        UARTResource(0, rx="Y20", tx="AE20",
+                     attrs=Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("i2c", 0,
+                 Subsignal("scl", Pins("AE30", dir="io")),
+                 Subsignal("sda", Pins("AF30", dir="io")),
+                 Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("ddram", 0,
```
Comment (whitequark) [inline nmigen_boards/genesys2.py:60]: This resource should be called `ddr` or `ddr2` or whatever standard it is on this board.

File: nmigen_boards/genesys2.py [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+                           DiffPairs(p="AB9", n="AC9", dir="o"),
+                           Attrs(IOSTANDARD="DIFF_SSTL15_DCI")),
+                 Subsignal("cke", Pins("AJ9", dir="o"),
+                           Attrs(IOSTANDARD="SSTL15_DCI")),
+                 Subsignal("odt", Pins("AK9", dir="o"),
+                           Attrs(IOSTANDARD="SSTL15_DCI")),
+                 Subsignal("reset", PinsN("AG5", dir="o"),
+                           Attrs(IOSTANDARD="SSTL15")),
+                 Attrs(SLEW="FAST", OUTPUT_IMPEDANCE="RDRV_40_40")),
+        Resource("audio", 0,
```
Comment (whitequark) [inline nmigen_boards/genesys2.py:96]: There's no single standardized "audio" interface. This looks like it's a combination of I2C, I2S, and something else? The resource should probably be split then.

File: nmigen_boards/genesys2.py [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+                 Subsignal("adc_sdata", Pins("AH19", dir="o")),
+                 Subsignal("dac_sdata", Pins("AJ19", dir="i")),
+                 Subsignal("bclk", Pins("AG18", dir="o")),
+                 Subsignal("lrclk", Pins("AJ18", dir="o")),
+                 Subsignal("mclk", Pins("AK19", dir="o")),
+                 Subsignal("scl", Pins("AE19", dir="io")),
+                 Subsignal("sda", Pins("AF18", dir="io")),
+                 Attrs(IOSTANDARD="LVCMOS18")),
+        Resource("oled", 0,
```
Comment (whitequark) [inline nmigen_boards/genesys2.py:105]: What interface does the display actually implement? SPI?

File: nmigen_boards/genesys2.py [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+                 Subsignal("data_cmd", PinsN("AC17", dir="o"),
+                           Attrs(IOSTANDARD="LVCMOS18")),
+                 Subsignal("reset", PinsN("AB17", dir="o"),
+                           Attrs(IOSTANDARD="LVCMOS18")),
+                 Subsignal("clk", Pins("AF17", dir="o"),
+                           Attrs(IOSTANDARD="LVCMOS18")),
+                 Subsignal("mosi", Pins("Y15", dir="o"),
+                           Attrs(IOSTANDARD="LVCMOS18")),
+                 Subsignal("vdd", PinsN("AG17", dir="o"),
+                           Attrs(IOSTANDARD="LVCMOS18")),
```
Comment (whitequark) [inline nmigen_boards/genesys2.py:115]: Is it actually powered by the FPGA pin? If not it should be something like `vdd_en`.

File: nmigen_boards/genesys2.py [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+                           Attrs(IOSTANDARD="DIFF_SSTL15_DCI",
+                                 ODT="RTT_40")),
+                 Subsignal("clk",
+                           DiffPairs(p="AB9", n="AC9", dir="o"),
+                           Attrs(IOSTANDARD="DIFF_SSTL15_DCI")),
+                 Subsignal("cke", Pins("AJ9", dir="o"),
+                           Attrs(IOSTANDARD="SSTL15_DCI")),
+                 Subsignal("odt", Pins("AK9", dir="o"),
+                           Attrs(IOSTANDARD="SSTL15_DCI")),
+                 Subsignal("reset", PinsN("AG5", dir="o"),
```
Comment (whitequark) [inline nmigen_boards/genesys2.py:93]: See the Atlys board for DDR pin naming.

File: nmigen_boards/genesys2.py [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+                 Subsignal("bclk", Pins("AG18", dir="o")),
+                 Subsignal("lrclk", Pins("AJ18", dir="o")),
+                 Subsignal("mclk", Pins("AK19", dir="o")),
+                 Subsignal("scl", Pins("AE19", dir="io")),
+                 Subsignal("sda", Pins("AF18", dir="io")),
+                 Attrs(IOSTANDARD="LVCMOS18")),
+        Resource("oled", 0,
+                 Subsignal("data_cmd", PinsN("AC17", dir="o"),
+                           Attrs(IOSTANDARD="LVCMOS18")),
+                 Subsignal("reset", PinsN("AB17", dir="o"),
```
Comment (whitequark) [inline nmigen_boards/genesys2.py:108]:
```suggestion
                 Subsignal("rst", PinsN("AB17", dir="o"),
```

File: nmigen_boards/genesys2.py [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+                 Subsignal("g", Pins("AJ23 AJ22 AH22 AK21 AJ21 AK23",
+                                     dir="o")),
+                 Subsignal("b", Pins("AH20 AG20 AF21 AK20 AG22", dir="o")),
+                 Subsignal("hsync", PinsN("AF20", dir="o")),
+                 Subsignal("vsync", PinsN("AG23", dir="o")),
+                 Attrs(IOSTANDARD="LVCMOS33")),
+        *SDCardResources(0, clk="R28", cmd="R29", dat0="R26", dat1="R30",
+                        dat2="P29", dat3="T30", cd="P28",
+                        attrs=Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("sd_card_reset", 0,
```
Comment (whitequark) [inline nmigen_boards/genesys2.py:151]:
```suggestion
        Resource("sd_card_rst", 0,
```

File: nmigen_boards/genesys2.py [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+                 Pins("AE24", dir="o"), Attrs(IOSTANDARD="LVCMOS33")),
+        Resource("usb_otg", 0,
```
Comment (whitequark) [inline nmigen_boards/genesys2.py:153]: This looks like an amalgamation of some interface, ULPI perhaps, and some power control pins, and needs to be split.

File: nmigen_boards/genesys2.py [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+                 Subsignal("stp", Pins("AA17", dir="o")),
+                 Subsignal("nxt", Pins("AA16", dir="i")),
+                 Subsignal("vbusoc", PinsN("AF16", dir="i")),
+                 Subsignal("reset", PinsN("AB14", dir="o")),
+                 Attrs(IOSTANDARD="LVCMOS18")),
+        Resource("eth_clocks", 0,
+                 Subsignal("tx_clk", Pins("AE10", dir="o")),
+                 Subsignal("rx_clk", Pins("AG10", dir="i")),
+                 Attrs(IOSTANDARD="LVCMOS18")),
+        Resource("eth", 0,
```
Comment (whitequark) [inline nmigen_boards/genesys2.py:167]: This looks like a specific kind of Ethernet interface (RGMII?)

Take a look at the Atlys board to see how an RGMII interface should look like.

File: nmigen_boards/genesys2.py (hunk 9 above)
Comment (peteut) [reply nmigen_boards/genesys2.py:60]: It's DDR3, I also rename `cke` to `clk_en` to follow the convention. And *DDR3_RESET_N* should become a dedicated resource.

Comment (peteut) [reply nmigen_boards/genesys2.py:60]: `reset` will be renamed to `rst` as well here.

Comment (peteut) [reply nmigen_boards/genesys2.py:60]: * `ras_n` → `ras` 
* `cas_n` → `cas`

File: nmigen_boards/genesys2.py (hunk 17 above)
Comment (peteut) [reply nmigen_boards/genesys2.py:167]: `eth` →`eth_rgmii`, subsignals as per `AtlysPlatform`.

File: nmigen_boards/genesys2.py (hunk 10 above)
Comment (peteut) [reply nmigen_boards/genesys2.py]: Split the resources:
* `audio_i2c`, ADAU1761 I2C, control port
* `audio_i2s`, ADAU1761 ADC/DAC, serial data ADC, DAC, [https://en.wikipedia.org/wiki/I²S](https://en.wikipedia.org/wiki/I%C2%B2S)
* `audio_clk`

File: nmigen_boards/genesys2.py (hunk 11 above)
Comment (peteut) [reply nmigen_boards/genesys2.py:101]: @whitequark it's an SPI interface w/o `cs` and `miso` pins. `SPIResource` expects those pins to be present. How should I tackle this?

**Comment by whitequark** [comment]
Thank you for your patience! All looks good to me now.

**Pull request merged**